x/dex/keeper/limitorder.go

```diff
@@ -0,0 +1,231 @@
+		return sdkerrors.Wrapf(types.ErrCannotWithdrawLimitOrder, "sharesOut is larger than shares Owned at the specified tick")
+	}
+
+	trancheUser.SharesCancelled = trancheUser.SharesCancelled.Add(sharesOut)
+
+	tranche.ReservesTokenIn = tranche.ReservesTokenIn.Sub(sharesOut)
+	return nil
+}
+
+func (tranche *LimitOrderTranche) WithdrawFilledLimitOrder(trancheUser *LimitOrderTrancheUser) sdk.Int {
+	amountFilled := tranche.priceOutToIn.MulInt(tranche.TotalTokenOut)
+	ratioFilled := amountFilled.QuoInt(tranche.TotalTokenIn)
+	maxAllowedToWithdraw := sdk.MinInt(
+		ratioFilled.MulInt(trancheUser.SharesOwned).TruncateInt(), // cannot withdraw more than what's been filled
+		trancheUser.SharesOwned.Sub(trancheUser.SharesCancelled),  // cannot withdraw more than what you own
+	)
+	amountOutTokenIn := maxAllowedToWithdraw.Sub(trancheUser.SharesWithdrawn)
+
+	amountOutTokenOut := tranche.priceInToOut.MulInt(amountOutTokenIn).TruncateInt()
+
+	trancheUser.SharesWithdrawn = maxAllowedToWithdraw
+
+	tranche.ReservesTokenOut = tranche.ReservesTokenOut.Sub(amountOutTokenOut)
+	return amountOutTokenOut
+}
+
+func (tranche *LimitOrderTranche) Swap() error {
+	// reservesTokenOut := &tranche.ReservesTokenIn
+	// fillTokenIn := &tranche.ReservesTokenOut
+	// totalTokenIn := &tranche.TotalTokenOut
+	// amountFilledTokenOut := priceInToOut.Mul(amountRemainingTokenIn).TruncateInt()
+	//
+	// if reservesTokenOut.LTE(amountFilledTokenOut) {
+	// amountOut = amountOut.Add(*reservesTokenOut)
+	// amountFilledTokenIn := priceOutToIn.MulInt(*reservesTokenOut)
+	// amountRemainingTokenIn = amountRemainingTokenIn.Sub(amountFilledTokenIn)
+	// *reservesTokenOut = sdk.ZeroInt()
+	// *fillTokenIn = fillTokenIn.Add(amountFilledTokenIn.TruncateInt())
+	// *totalTokenIn = totalTokenIn.Add(amountFilledTokenIn.TruncateInt())
+	// } else {
+	// amountOut = amountOut.Add(amountFilledTokenOut)
+	// *fillTokenIn = fillTokenIn.Add(amountRemainingTokenIn.TruncateInt())
+	// *totalTokenIn = totalTokenIn.Add(amountRemainingTokenIn.TruncateInt())
+	// *reservesTokenOut = reservesTokenOut.Sub(amountFilledTokenOut)
+	// amountRemainingTokenIn = sdk.ZeroDec()
+	// }
+	return nil
+}
+
+// TODO: possibly move this into the tranche object?
+type LimitOrderTrancheUser struct {
+	types.LimitOrderTrancheUser
+}
+
+func (loUser *LimitOrderTrancheUser) Save(ctx sdk.Context, k Keeper) {
+	k.SetLimitOrderTrancheUser(ctx, loUser.LimitOrderTrancheUser)
+}
+
+func (k Keeper) GetLimitOrderTrancheUserShares(goCtx context.Context, lo *LimitOrderTranche, user string) LimitOrderTrancheUser {
+	return LimitOrderTrancheUser{
+		k.GetOrInitLimitOrderTrancheUser(goCtx,
+			lo.LimitOrderTranche.PairId,
+			lo.LimitOrderTranche.TickIndex,
+			lo.LimitOrderTranche.TokenIn,
+			lo.LimitOrderTranche.TrancheIndex,
+			user),
+	}
+}
+
+func (k Keeper) GetCurrentPlaceTranche(ctx sdk.Context, tokenIn string, token0 string, pairId string, tick *types.Tick, placeTrancheIndex *uint64) (tranche *LimitOrderTranche, err error) {
```

Comment:
Put this into an LimitOrderBook domain object, not on the keeper. LimitOrderBook should manage all of the limit order tranches for a given tick-token combo.